File: common/src/main/java/net/polyv/android/player/common/ui/component/PLVMediaPlayerMoreLayoutDownloadActionView.java
package net.polyv.android.player.common.ui.component;

import static net.polyv.android.player.sdk.foundation.lang.PreconditionsKt.requireNotNull;
import static net.polyv.android.player.sdk.foundation.lang.StringsKt.format;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import net.polyv.android.player.common.R;
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider;
import net.polyv.android.player.common.modules.download.single.viewmodel.PLVMPDownloadItemViewModel;
import net.polyv.android.player.common.modules.download.single.viewmodel.viewstate.PLVMPDownloadItemViewState;
import net.polyv.android.player.common.ui.router.PLVMediaPlayerRouter;
import net.polyv.android.player.common.ui.router.RouterDestination;
import net.polyv.android.player.common.ui.router.RouterPayload;
import net.polyv.android.player.sdk.addon.download.common.model.vo.PLVMediaDownloadStatus;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberState;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberStateCompareResult;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

/**
 * @author Hoshiiro
 */
public class PLVMediaPlayerMoreLayoutDownloadActionView extends FrameLayout implements View.OnClickListener, View.OnLongClickListener {

    private ImageView downloadActionIv;
    private FrameLayout downloadProgressLayout;
    private ProgressBar downloadProgressBar;
    private TextView downloadProgressTv;
    private TextView downloadActionTv;

    private PLVMediaDownloadStatus downloadStatus = PLVMediaDownloadStatus.NOT_STARTED.INSTANCE;
    private float downloadProgress = 0;
    private boolean isVisible = true;

    public PLVMediaPlayerMoreLayoutDownloadActionView(@NonNull Context context) {
        super(context);
        init(null);
    }

    public PLVMediaPlayerMoreLayoutDownloadActionView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public PLVMediaPlayerMoreLayoutDownloadActionView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    private void init(@Nullable AttributeSet attrs) {
        LayoutInflater.from(getContext()).inflate(R.layout.plv_media_player_ui_component_more_download_action_layout, this);
        downloadActionIv = findViewById(R.id.plv_media_player_download_action_iv);
        downloadProgressLayout = findViewById(R.id.plv_media_player_download_progress_layout);
        downloadProgressBar = findViewById(R.id.plv_media_player_download_progress_bar);
        downloadProgressTv = findViewById(R.id.plv_media_player_download_progress_tv);
        downloadActionTv = findViewById(R.id.plv_media_player_download_action_tv);

        parseAttrs(attrs);

        setOnClickListener(this);
        setOnLongClickListener(this);
        onViewStateChanged();
    }

    private void parseAttrs(@Nullable AttributeSet attrs) {
        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.PLVMediaPlayerMoreLayoutDownloadActionView);
        int iconColor = typedArray.getColor(R.styleable.PLVMediaPlayerMoreLayoutDownloadActionView_plvIconTintNormal, Color.WHITE);
        int textColor = typedArray.getColor(R.styleable.PLVMediaPlayerMoreLayoutDownloadActionView_plvTextColorNormal, Color.WHITE);
        typedArray.recycle();
        downloadActionIv.setColorFilter(iconColor);
        downloadActionTv.setTextColor(textColor);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (isInEditMode()) return;

        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPDownloadItemViewModel.class)
                .getDownloadItem()
                .observeUntilViewDetached(this, new Function1<PLVMPDownloadItemViewState, Unit>() {
                    @Override
                    public Unit invoke(PLVMPDownloadItemViewState viewState) {
                        if (viewState == null) {
                            return null;
                        }
                        downloadStatus = viewState.getStatus();
                        downloadProgress = viewState.getProgress();
                        isVisible = viewState.isVisible();
                        onViewStateChanged();
                        return null;
                    }
                });
    }

    private void onViewStateChanged() {
        PLVRememberState.rememberStateOf(this, "updateDownloadStatus")
                .compareLastAndSet(downloadStatus)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult result) {
                        updateDownloadStatus();
                        return null;
                    }
                });

        PLVRememberState.rememberStateOf(this, "updateDownloadProgress")
                .compareLastAndSet(downloadProgress)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult result) {
                        updateDownloadProgress();
                        return null;
                    }
                });

        PLVRememberState.rememberStateOf(this, "updateVisibleState")
                .compareLastAndSet(isVisible)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult plvRememberStateCompareResult) {
                        updateVisibleState();
                        return null;
                    }
                });
    }

    private void updateDownloadStatus() {
        final String downloadText;
        final boolean isShowDownloadProgress;
        if (downloadStatus instanceof PLVMediaDownloadStatus.DOWNLOADING) {
            downloadText = getContext().getString(R.string.plv_media_player_ui_component_download_text_downloading);
            isShowDownloadProgress = true;
        } else if (downloadStatus instanceof PLVMediaDownloadStatus.WAITING) {
            downloadText = getContext().getString(R.string.plv_media_player_ui_component_download_text_waiting);
            isShowDownloadProgress = true;
        } else if (downloadStatus instanceof PLVMediaDownloadStatus.COMPLETED) {
            downloadText = getContext().getString(R.string.plv_media_player_ui_component_download_text_completed);
            isShowDownloadProgress = false;
        } else if (downloadStatus instanceof PLVMediaDownloadStatus.ERROR) {
            downloadText = getContext().getString(R.string.plv_media_player_ui_component_download_text_failed);
            isShowDownloadProgress = false;
        } else {
            downloadText = getContext().getString(R.string.plv_media_player_ui_component_download_text);
            isShowDownloadProgress = false;
        }
        downloadActionTv.setText(downloadText);
        downloadActionIv.setVisibility(isShowDownloadProgress ? GONE : VISIBLE);
        downloadProgressLayout.setVisibility(isShowDownloadProgress ? VISIBLE : GONE);
    }

    private void updateDownloadProgress() {
        int progressPercent = (int) (downloadProgress * 100);
        downloadProgressBar.setProgress(progressPercent);
        downloadProgressTv.setText(format("{}%", progressPercent));
    }

    private void updateVisibleState() {
        setVisibility(isVisible ? VISIBLE : GONE);
    }

    @Override
    public void onClick(View v) {
        if (downloadStatus instanceof PLVMediaDownloadStatus.NOT_STARTED
                || downloadStatus instanceof PLVMediaDownloadStatus.PAUSED
                || downloadStatus instanceof PLVMediaDownloadStatus.ERROR) {
            requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                    .get(PLVMPDownloadItemViewModel.class)
                    .startDownload();
        } else {
            gotoDownloadCenter();
        }
    }

    @Override
    public boolean onLongClick(View v) {
        gotoDownloadCenter();
        return true;
    }

    private void gotoDownloadCenter() {
        final boolean gotoDownloadingTab = downloadStatus instanceof PLVMediaDownloadStatus.PAUSED
                || downloadStatus instanceof PLVMediaDownloadStatus.WAITING
                || downloadStatus instanceof PLVMediaDownloadStatus.DOWNLOADING
                || downloadStatus instanceof PLVMediaDownloadStatus.ERROR;
        PLVMediaPlayerRouter.finish(RouterDestination.DownloadCenter.class);
        PLVMediaPlayerRouter.routerTo(getContext(),
                new RouterDestination.DownloadCenter(
                        new RouterPayload.DownloadCenterPayload(gotoDownloadingTab ? 1 : 0)
                )
        );
    }

}

File: common/src/main/java/net/polyv/android/player/common/ui/component/PLVMediaPlayerMoreLayoutDownloadActionView.kt
package net.polyv.android.player.common.ui.component

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import net.polyv.android.player.common.R
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider
import net.polyv.android.player.common.modules.download.single.viewmodel.PLVMPDownloadItemViewModel
import net.polyv.android.player.common.ui.router.PLVMediaPlayerRouter.routerTo
import net.polyv.android.player.common.ui.router.RouterDestination.DownloadCenter
import net.polyv.android.player.common.ui.router.RouterPayload.DownloadCenterPayload
import net.polyv.android.player.sdk.addon.download.common.model.vo.PLVMediaDownloadStatus
import net.polyv.android.player.sdk.addon.download.common.model.vo.PLVMediaDownloadStatus.COMPLETED
import net.polyv.android.player.sdk.addon.download.common.model.vo.PLVMediaDownloadStatus.DOWNLOADING
import net.polyv.android.player.sdk.addon.download.common.model.vo.PLVMediaDownloadStatus.NOT_STARTED
import net.polyv.android.player.sdk.addon.download.common.model.vo.PLVMediaDownloadStatus.PAUSED
import net.polyv.android.player.sdk.addon.download.common.model.vo.PLVMediaDownloadStatus.WAITING
import net.polyv.android.player.sdk.foundation.lang.PLVRememberState.Companion.rememberStateOf
import net.polyv.android.player.sdk.foundation.lang.format

/**
 * @author Hoshiiro
 */
class PLVMediaPlayerMoreLayoutDownloadActionView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr), View.OnClickListener, OnLongClickListener {
    private val downloadActionIv by lazy { findViewById<ImageView>(R.id.plv_media_player_download_action_iv) }
    private val downloadProgressLayout by lazy { findViewById<FrameLayout>(R.id.plv_media_player_download_progress_layout) }
    private val downloadProgressBar by lazy { findViewById<ProgressBar>(R.id.plv_media_player_download_progress_bar) }
    private val downloadProgressTv by lazy { findViewById<TextView>(R.id.plv_media_player_download_progress_tv) }
    private val downloadActionTv by lazy { findViewById<TextView>(R.id.plv_media_player_download_action_tv) }

    private var downloadStatus: PLVMediaDownloadStatus = NOT_STARTED
    private var downloadProgress = 0f
    private var isVisible = false

    init {
        LayoutInflater.from(context).inflate(R.layout.plv_media_player_ui_component_more_download_action_layout, this)

        parseAttrs(attrs)

        setOnClickListener(this)
        setOnLongClickListener(this)
        onViewStateChanged()
    }

    private fun parseAttrs(attrs: AttributeSet?) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.PLVMediaPlayerMoreLayoutDownloadActionView)
        val iconColor = typedArray.getColor(
            R.styleable.PLVMediaPlayerMoreLayoutDownloadActionView_plv_icon_tint_normal,
            Color.WHITE
        )
        val textColor = typedArray.getColor(
            R.styleable.PLVMediaPlayerMoreLayoutDownloadActionView_plv_text_color_normal,
            Color.WHITE
        )
        typedArray.recycle()
        downloadActionIv.setColorFilter(iconColor)
        downloadActionTv.setTextColor(textColor)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (isInEditMode) return

        PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
            .get<PLVMPDownloadItemViewModel>()
            .downloadItem
            .observeUntilViewDetached(this) { viewState ->
                if (viewState == null) {
                    return@observeUntilViewDetached
                }
                downloadStatus = viewState.status
                downloadProgress = viewState.progress
                isVisible = viewState.isVisible
                onViewStateChanged()
            }
    }

    private fun onViewStateChanged() {
        rememberStateOf("updateDownloadStatus")
            .compareLastAndSet(downloadStatus)
            .ifNotEquals {
                updateDownloadStatus()
            }

        rememberStateOf("updateDownloadProgress")
            .compareLastAndSet(downloadProgress)
            .ifNotEquals {
                updateDownloadProgress()
            }

        rememberStateOf("updateVisibleState")
            .compareLastAndSet(isVisible)
            .ifNotEquals {
                updateVisibleState()
            }
    }

    private fun updateDownloadStatus() {
        val downloadText: String?
        val isShowDownloadProgress: Boolean
        when (downloadStatus) {
            is DOWNLOADING -> {
                downloadText = context.getString(R.string.plv_media_player_ui_component_download_text_downloading)
                isShowDownloadProgress = true
            }

            is WAITING -> {
                downloadText = context.getString(R.string.plv_media_player_ui_component_download_text_waiting)
                isShowDownloadProgress = true
            }

            is COMPLETED -> {
                downloadText = context.getString(R.string.plv_media_player_ui_component_download_text_completed)
                isShowDownloadProgress = false
            }

            is PLVMediaDownloadStatus.ERROR -> {
                downloadText = context.getString(R.string.plv_media_player_ui_component_download_text_failed)
                isShowDownloadProgress = false
            }

            else -> {
                downloadText = context.getString(R.string.plv_media_player_ui_component_download_text)
                isShowDownloadProgress = false
            }
        }
        downloadActionTv.text = downloadText
        downloadActionIv.setVisibility(if (isShowDownloadProgress) GONE else VISIBLE)
        downloadProgressLayout.visibility = if (isShowDownloadProgress) VISIBLE else GONE
    }

    private fun updateDownloadProgress() {
        val progressPercent = (downloadProgress * 100).toInt()
        downloadProgressBar.progress = progressPercent
        downloadProgressTv.text = "{}%".format(progressPercent)
    }

    private fun updateVisibleState() {
        visibility = if (isVisible) VISIBLE else GONE
    }

    override fun onClick(v: View?) {
        if (downloadStatus is NOT_STARTED
            || downloadStatus is PAUSED
            || downloadStatus is PLVMediaDownloadStatus.ERROR
        ) {
            PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
                .get<PLVMPDownloadItemViewModel>()
                .startDownload()
        } else {
            gotoDownloadCenter()
        }
    }

    override fun onLongClick(v: View?): Boolean {
        gotoDownloadCenter()
        return true
    }

    private fun gotoDownloadCenter() {
        val gotoDownloadingTab = downloadStatus is PAUSED
                || downloadStatus is WAITING
                || downloadStatus is DOWNLOADING
                || downloadStatus is PLVMediaDownloadStatus.ERROR
        context.routerTo(
            DownloadCenter(
                DownloadCenterPayload(if (gotoDownloadingTab) 1 else 0)
            )
        )
    }
}
